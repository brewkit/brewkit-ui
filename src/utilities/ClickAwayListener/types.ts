import { ReactNode } from 'react';


export interface Props {

    /**
     * The content to be wrapped and not trigger the `onClickAway` hook.
     */
    children?: ReactNode,

    /**
     * The callback function to fire when the user clicks on an element not within the wrapped content.
     */
    onClickAway: () => void,

}
