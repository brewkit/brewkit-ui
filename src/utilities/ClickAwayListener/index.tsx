import React, { ReactElement, RefObject } from 'react';
import { Props } from './types';


/**
 * The `ClickAwayListener` component.
 */
function ClickAwayListener({
    onClickAway,
    children,
}: Props): ReactElement {


    const ref: RefObject<HTMLInputElement> = React.useRef(null);


    /**
     * Check to see if the click originated from a descendent of the wrapped content.
     */
    function handleClick(event: Event): void {
        const target = ref?.current?.parentNode?.firstChild;
        const isContained = target.contains(event.target as Node);
        if (!isContained) onClickAway();
    }


    React.useEffect(() => {
        document.addEventListener('click', handleClick);
        return ((): void => {
            document.removeEventListener('click', handleClick);
        });
    }, []);


    return (
        <React.Fragment>
            {children}
            <div ref={ref} style={{ display: 'none' }} />
        </React.Fragment>
    );


}


export default ClickAwayListener;
