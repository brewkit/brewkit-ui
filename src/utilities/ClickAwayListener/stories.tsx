import React, { ReactElement } from 'react';
import Label from '../../components/Label';
import ClickAwayListener from './index';
import Button from '../../components/Button';


export default {
    component: ClickAwayListener,
    parameters: {
        componentSubtitle: <Label color="success">Stable</Label>,
    },
    title: 'Utilities|ClickAwayListener',
};


export const General = (): ReactElement => {


    const [isLoading, setLoading] = React.useState(false);


    return (
        <ClickAwayListener onClickAway={(): void => setLoading(!isLoading)}>
            <Button isLoading={isLoading}>Click away from me!</Button>
        </ClickAwayListener>
    );


};
