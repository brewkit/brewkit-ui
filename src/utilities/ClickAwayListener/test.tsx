import React from 'react';
import renderer from 'react-test-renderer';
import ClickAwayListener from './index';


it('renders correctly', () => {
    const tree = renderer
        .create((
            <ClickAwayListener onClickAway={(): boolean => true}>
                <div>foo</div>
            </ClickAwayListener>
        ))
        .toJSON();
    expect(tree).toMatchSnapshot();
});
