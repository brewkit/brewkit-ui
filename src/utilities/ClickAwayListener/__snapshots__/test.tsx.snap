// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div>
    foo
  </div>,
  <div
    style={
      Object {
        "display": "none",
      }
    }
  />,
]
`;
